### ISSUE TITLE
Unify task-distribution sampling: TaskDistribution config powering both training + evaluation

### ISSUE THREAD

**Issue by mmcdermott** [payalchandak/EveryQuery#144]
> **Rewritten 2026-04-22** after design feedback from @mmcdermott. Previous draft treated "task distribution" as a config for the current scalar `(code, duration)` shape; this rewrite anchors on the recursive-grammar direction the repo is actually heading in (composite codes, event-bounded durations, explicit time windows) and on what is load-bearing for the NeurIPS submission. Previous draft is in the edit history.

## The problem we're solving

EveryQuery's task grammar is expanding along three concrete axes in the near-term, each of which breaks assumptions baked into today's two generators:

1. **Composite code tasks** — `ANY [A, B, C]`, `ALL [A, B, C]`, plus richer composition (`EXACT_SUBSEQUENCE`, `ORDERED_SET`, ...). A partial out-of-date PR exists and needs a principled home, not more flags on top of the flat shape.
2. **Event-bounded durations** — `duration = until the next event of code X`. Canonical tasks: in-hospital mortality (ends at DEATH-or-DISCHARGE), length-of-stay (ends at DISCHARGE). Not expressible as a scalar.
3. **Explicit time windows** — the window is `(start_event, end_event)`, not `(prediction_time, prediction_time + days)`. Canonical task: "did a negative event happen during an intervention period."

Today (`src/every_query/generate_tasks/sample_tasks.py`, `sample_evaluation_tasks.py`): `code` is a single string, `duration_days` is a `float32`, the window is always `(pred_time, pred_time + days)`. Each of the three extensions requires a different kind of structural change to each of these; adding them one-at-a-time on the current flat shape will accrete flags without a coherent center.

Task-type generalization (survival, regression, multiclass, different model heads — the old #129 direction) is a **separate future concern**, not what this umbrella is about.

## Vision — mirror `Task` and `TaskDistribution`

A `Task` is a named set of fields, each field a recursive expression tree:

```python
@dataclass(frozen=True)
class Task:
    code: CodeExpr       # CodeExpr = Leaf("A") | Any([CodeExpr, ...]) | All([CodeExpr, ...]) | ...
    time_range: TimeRange  # FixedDuration(30) | EventBounded(anchor, direction, fallback) | ExplicitWindow(start_expr, end_expr)
```

`TaskDistribution` **mirrors** `Task`: same field names, values replaced with distributions over trees:

```python
@dataclass(frozen=True)
class TaskDistribution:
    code: CodeExprDist   # LeafDist(metadata_file, filter, weight_col) | AnyDist(n_children, children_dist) | ...
    time_range: TimeRangeDist
```

Every realization is a `Task`; the space of realizations is a `TaskDistribution`. New grammar variants land on both sides symmetrically — you cannot add a sampler for `Any` without defining what `Any` means as a realized task.

Why this factoring:

- Composite-code sampling becomes a **recursive walk** over the tree — each `AnyDist` node draws `n_children` from its children distributions, which themselves may be `AnyDist` or `LeafDist`.
- Conditional distributions fall out naturally — a `time_range` distribution can read a column off the code's metadata (e.g., "if `clinical_category == oncology`, use categorical `[30, 90]`").
- Exclusions are tree-compared directly — no per-field special cases.

## Code attributes via a metadata parquet

Instead of inline weights + standalone excluded-codes lists, the code universe is **a parquet with attribute columns** — one row per code, columns like `weight`, `in_distribution`, `clinical_category`, `frequency`, `n_subjects`, etc.:

```
code              weight  in_distribution  clinical_category  frequency
ICD10//I10        1.0     true             cardiology         0.42
ICD10//C34        5.0     true             oncology           0.001
```

A `LeafDist` references any subset of columns for filtering / weighting:

```yaml
leaf:
  source: $PROCESSED/metadata/codes_with_attrs.parquet
  filter:
    all_of:
      - {column: in_distribution, op: equals, value: true}
      - {column: clinical_category, op: not_in, value: [oncology, admin]}
  weight_column: weight
```

Subsumes the previous umbrella's inline-weights concern + the previous `#147`'s explicit-exclusions list in one primitive. Metadata files are versionable, reusable across runs, and extensible (new columns = new filter axes, no YAML schema bumps).

## Task-level exclusions (separate primitive)

Distinct from metadata filtering. Metadata filtering shrinks the *support*; task-level exclusion cuts **specific realized tasks**, optionally with a radius covering continuous fields:

```yaml
exclusions:
  - task: {code: {leaf: DEATH}, time_range: {fixed_duration: 30}}
    radius: {time_range: {duration_days: 10}}    # also excludes DEATH at d ∈ [20, 40]
  - task: {code: {any: [HR_HIGH, HR_LOW]}, time_range: {fixed_duration: 7}}
    # exact match, no radius
```

Bubble-style exclusions (anchor_code + pre/post days from the previous `#147` draft) **subsume into event-bounded durations**: "no tasks post-DEATH" is just "use `EventBounded(anchor=DEATH, direction=forward)` for duration." One less primitive.

## Sampling as a pipeline

The sampling engine takes a list of **stages** and concatenates outputs:

```yaml
sampling:
  - mode: random                 # iid draws from the distribution
    n: 10_000
    distribution: base_distribution
  - mode: grid                   # sample each field independently, cartesian product
    n_per_field:
      code: all                  # enumerate every code matching distribution's filter
      time_range: 5              # 5 time_range draws per code
    distribution: focus_distribution
  - mode: local_radius           # from anchors, vary one continuous field
    anchor_tasks: [...]
    radius_per_field:
      time_range: [1, 7, 30, 90, 365]
# concatenate stage outputs
```

Three modes cover the real use-cases:

- **Random** — iid from the distribution (pretraining, broad eval coverage).
- **Grid** — fields sampled independently, cartesian product (detailed trade-off comparisons; today's `EQ_generate_evaluation_tasks` is this mode with `n_per_field.code = all`).
- **Local radius** — from a set of anchor tasks, vary one continuous field while fixing others. This is how you robustly capture per-`(code, duration)` monotonicity gradients without blowing up the grid.

Concatenation lets one config produce "10k random + full grid of 5 focus codes × 10 durations + local radius around 3 hero tasks."

## Child issues (post-rewrite)

**Foundation — all NeurIPS-blocking:**

- **#145** — `TaskDistribution` scaffold: typed mirror of `Task`, tree-recursive on fields, Hydra wiring, parquet schema evolution plan. Rewritten; now the *central* foundation issue.
- **#146** — Metadata-file-driven codes: parquet source + filter expression + weight column. Rewritten; replaces inline-weight surface.
- **#149** — Determinism audit. Light edit only — direction unchanged.
- **#150** — Rewire training-validity through the real path. Light edit only — direction unchanged.

**Grammar extensions — new issues:**

- **#154** — Composite code tasks: `Any`, `All`, later `ExactSubsequence`, `OrderedSet`. Partial stale PR is the starting point.
- **#155** — Event-bounded durations: `EventBounded(anchor_code, direction, fallback)`. Canonical mortality / LOS tasks.
- **#156** — Explicit time windows: `ExplicitWindow(start_event_expr, end_event_expr)`. Intervention-period evaluation.

**Orchestration — lower priority:**

- **#147** — Task-level exclusion (realized task + radius). Scope narrowed; bubbles subsumed by #155.
- **#148** — Full sampling pipeline (grid + random + local + concat). Today's dense eval grid is one mode; pipeline ships post-NeurIPS.

## NeurIPS critical path (~3 weeks)

Assuming the paper's headline tasks include event-bounded mortality-style durations.

**Week 1 — foundation.**
- **#145** scaffold
- **#146** metadata codes
- **#149** determinism (parallel)

**Week 2 — grammar + rewire.**
- **#155** event-bounded durations
- **#150** training-validity rewire
- **#154** composite codes — **blocking iff the paper has ANY/ALL headline tasks, else defer**

**Week 3 — paper runs + polish.**
- **#147** task-level exclusion (nice-to-have for OOD hold-out setup)
- Paper runs against the new pipeline

**Post-NeurIPS — full realization:**
- **#148** sampling pipeline (grid + random + local + concat)
- **#156** explicit time windows
- Conditional distributions (per-category duration overrides read from metadata)
- Tree-recursive composite operators beyond Any/All

### Fallback — narrow NeurIPS-only path

If three weeks is tight, the pragmatic alternative is to ship **just event-bounded durations + metadata filter inside the current generate_tasks code path**, without the full scaffold refactor. That gets the paper's headline tasks working without the Task-↔-TaskDistribution rewiring. Do the full abstraction post-NeurIPS as a separate landing.

Call this **Option B**. It is less principled but ships faster and leaves fewer things to go wrong during the submission crunch. Recommend Option A (full foundation) *iff* there's confidence in the week-1 scaffold landing cleanly; otherwise Option B.

## What changed vs. the previous draft

- **Grammar direction corrected** — recursive tree on existing fields, not new task types.
- **Task ↔ TaskDistribution mirror** as the central abstraction (didn't exist in the previous draft).
- **Metadata-file-driven codes** — replaces inline per-code weights + exclusion lists.
- **Task-level exclusion with radius** — new primitive. Bubbles subsumed by event-bounded durations.
- **Sampling pipeline** (grid / random / local / concat) — one config composes all modes.
- **Three grammar extensions as first-class child issues** (composite, event-bounded, explicit-window) — were not present before.
- **NeurIPS path + fallback** — explicit triage and a narrow Option B.

## Non-goals (unchanged)

- Label semantics unchanged (`TaskQuerySchema.boolean_value` nullable).
- New task types out of scope — separate concern (#129).
- `sample_codes/` stays in `paper_experiments/` per #85.

## Open design questions

- **Parquet schema evolution.** `TaskQuerySchema` is flat today. Under the tree grammar, what does the on-disk shape look like? Three candidates: (1) JSON-encoded Task column (simple, opaque); (2) widen flat schema with nullable `anchor_code`, `anchor_direction`, `composite_kind`, ...; (3) parquet nested struct. Recommend (3) — polars-native, query-friendly. Detail defers to #145.
- **Train-time vs eval-time distribution coupling.** Pin the *universe* (metadata file + filter) to the checkpoint; let the sampling pipeline vary per-run. Thoughts?
- **Stale composite-codes PR.** How salvageable? Drives whether #154 starts from zero or evolves the existing partial work.

## Questions for @payalchandak

- Which of the three grammar extensions (composite, event-bounded, explicit-window) are in the NeurIPS paper's headline tasks? Drives priority of #154/B/C.
- Pointer to the stale composite-codes PR?
- For mortality-style tasks, preferred event-bounded shape: "until first DEATH or DISCHARGE, fallback 30d"? Something else?
- Is the paper's ablation axis driven by clinical-category groupings that would live in the metadata file?
- Option A (full foundation) vs Option B (narrow NeurIPS-only) — which are you comfortable with given the timeline?
